### PR TITLE
Add missing download and token parameter in File interface

### LINKED ISSUE #14870: Can't download own file
### Describe the Bug

As a admin (full access), I can't download a file I uploaded.
Even If I refresh the page in a collection.

### To Reproduce

Create a collection, add a file field.
Upload a random file to the collection in this field.
Try Download it.
download-error

### Errors Shown

403

### What version of Directus are you using?

9.14.5

### What version of Node.js are you using?

6.4.16

### What database are you using?

mysql

### What browser are you using?

edge

### How are you deploying Directus?

localy

### PR DESCRIPTION
**Pull request by JonathanSchndr**
## Description

Add missing download and token parameter to the file download action

Fixes #14870

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated

### CODE REVIEW

File: app/src/interfaces/file/file.vue
Comment (azrikahar) [inline app/src/interfaces/file/file.vue:49]: I believe we should still keep the `download` prop here, since it's meant to be passed down to the `<a>` tag/component in `<v-list-item>` for the download filename: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download as the "Filename (Download)" field is customizable in each file.

**Review by azrikahar — CHANGES_REQUESTED** [review]

**Review by nervusdm — APPROVED** [review]

File: app/src/interfaces/file/file.vue
Comment (JonathanSchndr) [reply app/src/interfaces/file/file.vue:49]: Yes, I had also thought at first, but then I looked in the files-interface and took this logic, maybe this must be adjusted there?

Comment (JonathanSchndr) [reply app/src/interfaces/file/file.vue]: download name is added in interface file

Comment (azrikahar) [reply app/src/interfaces/file/file.vue]: > Yes, I had also thought at first, but then I looked in the files-interface and took this logic, maybe this must be adjusted there?

Makes sense! Let's add the download prop in `files` interface as well. Likely was just a missed detail when adding download link to `files` interface in ##12220.

**Comment by JonathanSchndr** [comment]
The changes are in:

- [x] Add header download action to file drawer (like files drawer)
- [x] Add for all downloads :download attr
- [x] Update file libary save_alt icon to new download icon and update to href link

File: app/src/interfaces/files/files.vue
Comment (rijkvanzanten) [inline app/src/interfaces/files/files.vue:218]: `directus_files_id` can't be hardcoded, as the name of the field used in the relationship isn't guaranteed. Instead, we should update the `computed` `fields` above to include the correct field name from the `relationInfo` ref 👍🏻

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

File: app/src/interfaces/files/files.vue
Comment (JonathanSchndr) [reply app/src/interfaces/files/files.vue:218]: I hope I had understood you correctly, this would be my adjustment:

```
const fields = computed(() =>
	adjustFieldsForDisplays(
		[...getFieldsFromTemplate(templateWithDefaults.value), `${relationInfo.value?.relation.field}.filename_download`],
		relationInfo.value?.junctionCollection.collection ?? ''
	)
);
```

if that is correct, i will commit the change.

Comment (rijkvanzanten) [reply app/src/interfaces/files/files.vue:218]: LGTM!

File: app/src/interfaces/file/file.vue
Comment (azrikahar) [inline app/src/interfaces/file/file.vue:49]: I think there's an oddity with the recent merge from main here. The removal of `downloadURL` which was a recent fix in #15126 results in an unused computed `downloadURL` now. I see that they're functionally the same aside from `downloadURL` not having the `download` query param, but I do personally prefer the use of computed value. @licitdev thoughts on this?

File: app/src/modules/files/routes/item.vue
Comment (azrikahar) [inline app/src/modules/files/routes/item.vue:352]: Similar to the first comment, perhaps we can "repurpose" this as a computed `downloadFileURL` or `downloadURL` instead of the direct addTokenToURL usage in template? Totally get that how using a computed can feel like an extra step, but just thinking aloud from a code consistency perspective. I'll defer to @rijkvanzanten for this 😉

File: app/src/interfaces/file/file.vue
Comment (azrikahar) [inline app/src/interfaces/file/file.vue:161]: Let's remove this line. Likely the same merge from main issue here, as line 169 now already has the same import statement, thus failing the CI lint check for this PR: https://github.com/directus/directus/runs/7919553314?check_suite_focus=true#step:3:201

File: app/src/interfaces/file/file.vue
Comment (licitdev) [reply app/src/interfaces/file/file.vue:49]: I believe that issue #14870 has been fixed alongside #15126. 🤔 

While the `?download` query param isn't necessary to allow for the download in app, it's a nice addition to ensure that the download is forced when users copy the download link. 👍 

I'd personally prefer the use of computed value as well.

File: app/src/modules/files/routes/item.vue
Comment (rijkvanzanten) [reply app/src/modules/files/routes/item.vue:352]: Yeah this should be a computed instead of doing it in-template 👍🏻

File: app/src/composables/use-asset-url.ts
Comment (azrikahar) [inline app/src/composables/use-asset-url.ts:9]: This feels rather odd as a composable, at least with the current way it's directly returning the value of a computed property 🤔 Not sure if this is idiomatic in terms of composables.

I think it should just be a util function, or alternatively as an exported function alongside `addTokenToURL` directly within the `@/api` file.

Comment (JonathanSchndr) [reply app/src/composables/use-asset-url.ts:9]: Hey azrikahar, Thanks for your feedback! I had created it as Composables, since it was going to be a Computed. Utlitis would be here a normal JavaScript function without caching. Gladly I can change this of course!

Comment (azrikahar) [reply app/src/composables/use-asset-url.ts:9]: Makes sense! let's go with the utility route 👍

File: app/src/utils/get-asset-url.ts
Comment (rijkvanzanten) [inline app/src/utils/get-asset-url.ts:5]: Lets use a URL object to make these modifications, and add some unit tests 🙂

Comment (JonathanSchndr) [reply app/src/utils/get-asset-url.ts]: unit test writing is very fun :D

**Review by licitdev — CHANGES_REQUESTED** [review]
Tested the `file` and `files` working well, with the download working when the link is copied.

@JonathanSchndr The use of `getAssetUrl` should also be propagated to the image interface. https://github.com/directus/directus/blob/6e222c9226402d17ed34c5844b71838883b71a1e/app/src/interfaces/file-image/file-image.vue#L145-L148

**Comment by JonathanSchndr** [comment]
> Tested the `file` and `files` working well, with the download working when the link is copied.
> 
> @JonathanSchndr The use of `getAssetUrl` should also be propagated to the image interface.
> 
> https://github.com/directus/directus/blob/6e222c9226402d17ed34c5844b71838883b71a1e/app/src/interfaces/file-image/file-image.vue#L145-L148

@licitdev - Thanks for the testing. I add this here as well :) I am very happy to see my feature getting so much new uses. Thank you !

**Comment by licitdev** [comment]
@JonathanSchndr Thank you for the fast updates! 

The following merge in https://github.com/directus/directus/commit/e68e45bafa87ab05226484cd5c7ce4844643b102 led to an error when opening the `v-drawer` in the `files` interface. 🤔 

image

@Nitwel How should it be resolved?

File: app/src/interfaces/files/files.vue
Comment (Nitwel) [inline app/src/interfaces/files/files.vue:325]:
```suggestion
	const junctionField = relationInfo.value.junctionField.field;
	const relationPkField = relationInfo.value.relatedPrimaryKeyField.field;

	return displayItems.value.find(item => get(item, [junctionField, relationPkField]))?.directus_files_id?.filename_download
```

**Review by Nitwel — CHANGES_REQUESTED** [review]
This is the proper way to get the `filename_download` now.

**Review by licitdev — CHANGES_REQUESTED** [review]
@JonathanSchndr Works well after merging with main! Let's add the download button to the `file-image` interface to be consistent with the `file` interface. 👍 

https://github.com/directus/directus/blob/95fab27f22eca9c7ee519272e77dc9e0de9da6ed/app/src/interfaces/file/file.vue#L90-L94

**Review by licitdev — APPROVED** [review]

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**
